catalyst-gateway/bin/src/service/utilities/middleware/schema_validation.rs

```diff
@@ -0,0 +1,58 @@
+//! Middleware to verify the status of the last DB schema version validation.
+//!
+//! If a mismatch is detected, the middleware returns an error with `ServiceUnavailable`
+//! status code (503). Otherwise, the middleware calls and returns the wrapped endpoint's
+//! response.
+//!
+//! This middleware checks the `State.schema_version_status` value, if it is Ok,
+//! the wrapped endpoint is called and its response is returned.
+
+use std::sync::Arc;
+
+use poem::{web::Data, Endpoint, EndpointExt, Middleware, Request, Result};
+
+use crate::{
+    service::common::responses::resp_5xx::ServiceUnavailable,
+    state::{SchemaVersionStatus, State},
+};
+
+/// A middleware that raises an error  with `ServiceUnavailable` and 503 status code
+/// if a DB schema version mismatch is found the existing `State`.
+pub(crate) struct SchemaVersionValidation;
+
+impl<E: Endpoint> Middleware<E> for SchemaVersionValidation {
+    type Output = SchemaVersionValidationImpl<E>;
+
+    fn transform(&self, ep: E) -> Self::Output {
+        SchemaVersionValidationImpl { ep }
+    }
+}
+
+/// The new endpoint type generated by the `SchemaVersionValidation`.
+pub(crate) struct SchemaVersionValidationImpl<E> {
+    /// Endpoint wrapped by the middleware.
+    ep: E,
+}
+
+#[poem::async_trait]
+impl<E: Endpoint> Endpoint for SchemaVersionValidationImpl<E> {
+    type Output = E::Output;
+
+    async fn call(&self, req: Request) -> Result<Self::Output> {
+        let req_data: Option<&Data<&Arc<State>>> = req.data();
+        if let Some(state) = req_data {
+            if state.is_schema_version_status(&SchemaVersionStatus::Mismatch) {
```

Comment:
what is the purpose to compare current schema version status with previous one at this place ?
Would be better to add some comments about it, seems not too obvious

Reply:
Added a comment there. Thanks.